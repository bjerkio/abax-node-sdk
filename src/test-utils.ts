import { MockAgent, type Interceptable, setGlobalDispatcher } from 'undici';
import { AbaxClient } from './abax-client';

export function initialiseClientAndMockPool(): {
  client: AbaxClient;
  mockPool: Interceptable;
} {
  const mockAgent = new MockAgent();
  mockAgent.disableNetConnect();
  setGlobalDispatcher(mockAgent);

  const mockPool = mockAgent.get('https://api.abax.cloud');

  const client = new AbaxClient({
    baseUrl: 'production',
    apiKey: 'kanonball!',
  });

  return { client, mockPool };
}
