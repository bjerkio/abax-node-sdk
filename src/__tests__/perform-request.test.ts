import { describe, expect, it } from 'vitest';
import { initialiseClientAndMockPool } from '../test-utils';

describe('perform request', () => {
  it.skip('Should back off and retry on 429', () => {
    const { client, mockPool } = initialiseClientAndMockPool();

    mockPool
      .intercept({
        path: '/v2/equipment/90c5b884d50b46afa6eb3aa14792f782',
        method: 'GET',
      })
      .reply(429, {})
      .times(3);

    const equipment = client.getEquipment({
      id: '90c5b884d50b46afa6eb3aa14792f782',
    });

    expect(equipment).rejects.toMatchSnapshot();
  });

  it('Should throw rejected error on 403', () => {
    const { client, mockPool } = initialiseClientAndMockPool();

    mockPool
      .intercept({
        path: '/v2/equipment/90c5b884d50b46afa6eb3aa14792f782',
        method: 'GET',
      })
      .reply(403, {
        error: 'Forbidden',
        message: 'Access denied',
      })
      .times(1);

    const equipment = client.getEquipment({
      id: '90c5b884d50b46afa6eb3aa14792f782',
    });

    expect(equipment).rejects.toMatchSnapshot();
  });

  it('Should throw unauthorized error on 401', () => {
    const { client, mockPool } = initialiseClientAndMockPool();

    mockPool
      .intercept({
        path: '/v2/equipment/90c5b884d50b46afa6eb3aa14792f782',
        method: 'GET',
      })
      .reply(401, {
        error: 'Unauthorized',
        message: 'Invalid token',
      })
      .times(1);

    const equipment = client.getEquipment({
      id: '90c5b884d50b46afa6eb3aa14792f782',
    });

    expect(equipment).rejects.toMatchSnapshot();
  });

  it('Should throw not found error on 404 with message', () => {
    const { client, mockPool } = initialiseClientAndMockPool();

    mockPool
      .intercept({
        path: '/v2/equipment/90c5b884d50b46afa6eb3aa14792f782',
        method: 'GET',
      })
      .reply(404, {
        error: 'Not Found',
        message: 'Equipment by id not found',
      })
      .times(1);

    const equipment = client.getEquipment({
      id: '90c5b884d50b46afa6eb3aa14792f782',
    });

    expect(equipment).rejects.toMatchSnapshot();
  });

  it('should pass through error message on 400', () => {
    const { client, mockPool } = initialiseClientAndMockPool();

    mockPool
      .intercept({
        path: '/v2/equipment/90c5b884d50b46afa6eb3aa14792f782',
        method: 'GET',
      })
      .reply(400, {
        error: 'Bad Request',
        message: 'Equipment id is invalid',
      })
      .times(1);

    const equipment = client.getEquipment({
      id: '90c5b884d50b46afa6eb3aa14792f782',
    });

    expect(equipment).rejects.toMatchSnapshot();
  });

  const errorCodes = [400, 401, 403, 404];
  errorCodes.forEach(errorCode => {
    it(`Should handle ${errorCode} error without message`, () => {
      const { client, mockPool } = initialiseClientAndMockPool();

      mockPool
        .intercept({
          path: '/v2/equipment/90c5b884d50b46afa6eb3aa14792f782',
          method: 'GET',
        })
        .reply(errorCode, {})
        .times(1);

      const equipment = client.getEquipment({
        id: '90c5b884d50b46afa6eb3aa14792f782',
      });

      expect(equipment).rejects.toMatchSnapshot();
    });
  });
});
