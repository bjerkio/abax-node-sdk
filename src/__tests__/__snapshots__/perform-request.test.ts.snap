// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`perform request > Should handle 400 error without message 1`] = `[Error: Request was rejected: {}]`;

exports[`perform request > Should handle 401 error without message 1`] = `[Error: Request was unauthorized: {}]`;

exports[`perform request > Should handle 403 error without message 1`] = `[Error: Request was forbidden: {}]`;

exports[`perform request > Should handle 404 error without message 1`] = `[Error: Not able to perform request]`;

exports[`perform request > Should throw not found error on 404 with message 1`] = `[Error: Not able to perform request]`;

exports[`perform request > Should throw rejected error on 403 1`] = `[Error: Request was forbidden: {"error":"Forbidden","message":"Access denied"}]`;

exports[`perform request > Should throw unauthorized error on 401 1`] = `[Error: Request was unauthorized: {"error":"Unauthorized","message":"Invalid token"}]`;

exports[`perform request > should pass through error message on 400 1`] = `[Error: Request was rejected: {"error":"Bad Request","message":"Equipment id is invalid"}]`;
